man/man1/evolvotron_render.1
.TH EVOLVOTRON_RENDER 1 "16 Oct 2009" "www.timday.com" "Evolvotron"

.SH NAME
evolvotron_render \- Render an evolvotron function tree to an image.

.SH SYNOPSIS
evolvotron_render
[options]
.I imagefile.[png|ppm]

.SH DESCRIPTION

.B evolvotron_render 
reads an evolvotron image function from its
standard input and renders it to an image in the file specified
(suffix determines type, defaults to ppm if not recognised).

Image functions can be obtained by saving them from the
evolvotron application, or using evolvotron_mutate.

See the evolvotron manual (accessible from the evolvotron
application's Help menu) for more information on image functions.

.SH COMMAND-LINE OPTIONS

.TP 0.5i
.B \-f, \-\-frames
.I frames
Generates multi-frame animations.  .fnnnnnn is inserted into
the specified filename (before the filetype suffix, if any).
You can use this on functions which weren't evolved in animation mode,
but there's no guarantee they have any interesting time/z variation.

.TP 0.5i
.B \-h, \-\-help
Display a summary of command-line options and exit.

.TP 0.5i
.B \-j, \-\-jitter
Enable sample jittering.

.TP 0.5i
.B \-m, \-\-multisample
.I multisample
Enables antialiased rendering.
This specifies the size of the sub-pixel sampling grid,
so 1 provides the default one-sample-per-pixel behaviour,
while 4 provides 16 samples per pixel on a 4x4 grid.
Unlike the main evolvotron application, there is no upper limit,
but of course rendering time increases as the square of this number.

.TP 0.5i
.B \-o, \-\-output
.I imagefile.[ppm|png]
This option is an alternative to specifying the output filename as a positional argument.

.TP 0.5i
.B \-s, \-\-size
.I widthxheight
Specify resolution of output image.
Defaults to 512x512.

.TP 0.5i
.B \-v, \-\-verbose
Verbose mode; useful for monitoring progress of large renders.

.SH EXAMPLES

evolvotron_mutate \-g | evolvotron_render \-s 1024x1024 function.ppm

.SH AUTHOR
.B evolvotron_render
was written by Tim Day (www.timday.com) and is released
under the conditions of the GNU General Public License.
See the file LICENSE supplied with the source code for details.

.SH SEE ALSO

evolvotron(1), evolvotron_mutate(1)


NEWS
[Most recent at top]

Release 0.7.1:
 - Uses `CONFIG += ordered` in `main.pro`; enables parallelised make.
 - Fold in typos fixed by Debian downstream.
 - tarball on Sourceforge files.

Release 0.7.0:
 - Got to compile vs Qt5.9.1 on Debian Stretch & Jessie; preparing for Qt5-only on Buster.
 - Get rid of precompiled header use (more trouble than it's worth these days).
 - Upgrade std::auto_ptr use to std::unique_ptr - cleans up compiler's deprecated warning spew.

From release 0.6.3:
 - Version to 0.6.4

From release 0.6.2:
 - Version to 0.6.3
 - Files derived from USAGE (evolvotron.html and usage_text.h) are now under
   svn control and no longer need to be generated as part of the build of a
   fresh checkout.
 - Migrated to new SourceForge platform.
   Needed: svn switch --relocate https://evolvotron.svn.sourceforge.net/svnroot/evolvotron "svn+ssh://timday@svn.code.sf.net/p/evolvotron/code/"
   
From release 0.6.1:
 - Version to 0.6.2
 - CHANGES renamed to NEWS
 - License boilerplate managed using "headache"
 - Add some "tartan" inspired functions.
 - Add pixel/voxel/hex grid quantizing functions.
 - Port command line options to boost::program_options,
   rationalize commandline options.
 - Fix to evolvotron_mutate from Pedro Gimeno.
 - Update mkdeb script to work on squeeze
 - Functions to be loaded on startup can be specified on the commandline.
 - Add option to shuffle startup functions.
 
From release 0.6.0:
 - Version to 0.6.1
 - Fix bug in saved functions: quotes missing around version attribute

From release 0.5.1:
 - Version to 0.6.0
 - Switch to Qt4.
 - (Qt4 related) keyboard shortcuts rationalized
   many removed, ctrl-modifiers dropped
 - (Qt4 related) image displays now have a frame around them
 - (Qt4 related) Remove non-standard Qt animated PNG support
 - evolvotron_match removed
 - Platform specific code (e.g number of processors identification)
   isolated in libevolvotron/platform_specific.[h|cpp]
   and selected by PLATFORM_LINUX or PLATFORM_BSD
   (Fink Qt's __DARWIN_X11__ also detected and used to select PLATFORM_BSD).
 - Repository migrated to svn

From release 0.5.0:
 - Version to 0.5.1
 - Replace not often used frontside heating/cooling buttons with autocooling mechanism.
 - Separate tabs for info & license on "About" dialog.
 - Better formatting of function properties dialog, and also shows XML function tree.
 - Dialog with rendering controls
 - Multisampling (antialiasing) options
 - -a/-j/-s command line options for autocool/jitter/multisample
 - evolvotron_render adds -j/-s to control jitter & multisampling
 - Option (-E) of separate thread pool (with different nices: -n/-N) for enlargements.
   Task totals for each reported separately.
 - Enlargement and solo renderings are fragmented into multiple smaller
   jobs to better utilize multiprocessor systems.
 - Generally better use of ref-counted QImage and QPixmap.
 - Middle-mouse drag adjustments flicker less.

From release 0.4.0:
 - Version to 0.5.0
 - Lists of functions in alphabetic order.
 - New FunctionTop always at top level.
 - New functions: Frieze groups.
 - Constantness of functions more conservative (less overrides) and hopefully less buggy.
 - Final colour space transform range more constrained to keep colours sane.
 - More sensible mutation of top level spatial transform (less arbitrary shearing).
 - Note that above changes mean old saved functions will appear different.
 - New mutation parameter: probability of a function's parameters being reset.
 - Code clean up:
   - More std::auto_ptr, boost::shared_ptr, boost::ptr_vector.  Code is now delete free.
   - Get rid of hacked static function registration; uses auto-generated register_all_functions.
 - Compute threads use QWaitCondition, not polling.
 - Default number of threads is number of processors.
 - Compute threads run at lower priority.
 - -n <niceness> command line option to control compute thread priority.
   (NB assumes Linux NPTL's non-Posix compliant behaviour of threads NOT sharing a common nice value;
   see "man pthreads" on Linux.)
 - About dialog displays number of threads.
 - Use precompiled headers (build time 2m45s down to 1m45s).
 - code_review script
 - Set an icon on various dialogs and main frame.
 - Add -D (function debug) massively weighting in favour of FunctionNoiseOneChannel.  Useful with -x/-X.
 - Believed to fix Debian bug #436231 and gentoo #144160
 - MutatableImage has serial number (fix problems caused by pointer value being wrongly used as equivalent)

From release 0.3.1:
 - Version to 0.4.0
 - Lock state moved to mutatable image so is stored in undo history
 - Mandelbrot & Julia set only in x,y plane.  Julia set uses fixed parameters.
 - Add Juliabrot.
 - Fix "Randomize" button on "All" functions tab.
 - "Favourite" function dialog provides access to -x/-X command-line functionality
 - Undo after reset works better
 - Fix broken FunctionGradient.
   Replace with GradientOfMagnitude, DivergenceOfMagnitude, Divergence, Curl
 - Migrate float->double via typedef real
 - Always builds with fullscreen option.

From release 0.3.0:
 - Version to 0.3.1
 - Fix for gcc 3.4
 - ./configure made (slightly) more robust

From release 0.2.3:
 - Version to 0.3.0
 - Get rid of troublesome system call to ../VERSION 
   in common.pro and pass settings from configure script instead
 - Add not-very-useful .qt-mng image file output format
 - Add function stats to Properties dialog
 - Add function simplification (more useful for debugging than anything)
 - Don't use True and False in doc-building python script
   (seems to be quite new thing, not in 2.1.x ?).  Use 0 and 1 instead.
 - Internal changes
 - Add pkg/debian .deb builder script (still under development)
 - Fix buggy Y rotation matrix
 - Eliminate template-tastic Tuple and Matrix classes.
   Fix XYZ and Transform to not require them.
 - evolvotron_mutate - -g option more likely to generate insteresting functions
 - -linz and -spheremap options where appropriate

evolvotron_render/evolvotron_render.cpp
/**************************************************************************/
/*  Copyright 2012 Tim Day                                                */
/*                                                                        */
/*  This file is part of Evolvotron                                       */
/*                                                                        */
/*  Evolvotron is free software: you can redistribute it and/or modify    */
/*  it under the terms of the GNU General Public License as published by  */
/*  the Free Software Foundation, either version 3 of the License, or     */
/*  (at your option) any later version.                                   */
/*                                                                        */
/*  Evolvotron is distributed in the hope that it will be useful,         */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         */
/*  GNU General Public License for more details.                          */
/*                                                                        */
/*  You should have received a copy of the GNU General Public License     */
/*  along with Evolvotron.  If not, see <http://www.gnu.org/licenses/>.   */
/**************************************************************************/

/*! \file
  \brief Standalone renderer for evolvotron function files.
*/

#include "function_registry.h"
#include "mutatable_image.h"
#include "random.h"

#include <boost/program_options.hpp>

//! Application code
int main(int argc,char* argv[])
{
  {
    uint frames;
    bool help;
    bool jitter;
    int multisample;
    std::string output_filename;
    std::string size;
    bool verbose;
    
    boost::program_options::options_description options_desc("Options");
    boost::program_options::positional_options_description pos_options_desc;
    {
      using namespace boost::program_options;
      options_desc.add_options()
	("frames,f"     ,value<uint>(&frames)->default_value(1)    ,"Frames in an animation")
	("help,h"       ,bool_switch(&help)                        ,"Print command-line options help message and exit")
	("jitter,j"     ,bool_switch(&jitter)                      ,"Enable rendering jitter")
	("multisample,m",value<int>(&multisample)->default_value(1),"Multisampling grid (NxN)")
	("output,o"     ,value<std::string>(&output_filename)      ,"Output filename (.png or .ppm suffix).  (Or use first positional argument.)")
	("size,s"       ,value<std::string>(&size)->default_value("512x515"),"Generated image size")
	("verbose,v"    ,bool_switch(&verbose)                     ,"Log some details to stderr")
	;
      pos_options_desc.add("output",1);
    }

    boost::program_options::variables_map options;
    boost::program_options::store
      (
       boost::program_options::command_line_parser(argc,argv)
       .options(options_desc).positional(pos_options_desc).run()
       ,options
       );
    boost::program_options::notify(options);
    
    if (help)
      {
	std::cerr << options_desc;
	return 0;
      }

    if (verbose)
      std::clog.rdbuf(std::cerr.rdbuf());
    else
      std::clog.rdbuf(sink_ostream.rdbuf());

    //! \todo Could be done better maybe (set 'x' as input separator)
    const std::string::size_type p=size.find("x");
    if (p==std::string::npos || p==0 || p==size.size()-1)
      {
	std::cerr << "--size option argument isn't in <width>x<height> format\n";
	return 1;
      }
    else
      {
	size[p]=' ';
      }
    int width=512;
    int height=512;
    std::stringstream(size) >> width >> height;
    
    if (frames<1)
      {
	std::cerr << "Must specify at least 1 frame (option: -f <frames>)\n";
	return 1;
      }

    if (output_filename.empty())
      {
	std::cerr << "Must specify an output filename\n";
	return 1;
      }
    
    FunctionRegistry function_registry;
    
    std::string report;
    const boost::shared_ptr<const MutatableImage> imagefn(MutatableImage::load_function(function_registry,std::cin,report));

    if (imagefn.get()==0)
      {
	std::cerr << "evolvotron_render: Error: Function not loaded due to errors:\n" << report;
	return 1;
      }
    else if (!report.empty())
      {
	std::cerr << "evolvotron_render: Warning: Function loaded with warnings:\n" << report;
      }

    // Seed value pretty unimportant; only used for sample jitter.
    Random01 r01(23);

    for (uint frame=0;frame<frames;frame++)
      {
	std::vector<uint> image_data;
	image_data.reserve(width*height);
  
	uint pixels=0;
	uint report=1;
	const uint reports=20;
	for (int row=0;row<height;row++)
	  for (int col=0;col<width;col++)
	    {
	      const XYZ v(imagefn->sampling_coordinate(col,row,frame,width,height,frames));
	    
	      const XYZ colour(imagefn->get_rgb(col,row,frame,width,height,frames,(jitter ? &r01 : 0),multisample));
	    
	      const uint col0=lrint(clamped(colour.x(),0.0,255.0));
	      const uint col1=lrint(clamped(colour.y(),0.0,255.0));
	      const uint col2=lrint(clamped(colour.z(),0.0,255.0));

	      image_data.push_back(((col0<<16)|(col1<<8)|(col2)));

	      pixels++;
	      if (pixels>=(report*width*height)/reports)
		{
		  std::clog << "[" << (100*report)/reports << "%]";
		  report++;
		}
	    }
	std::clog << "\n";

	{
	  //! \todo If filename is "-", write PPM to stdout (QImage save only supports write-to-a-filenames though)
	  QString save_filename(QString::fromLocal8Bit(output_filename.c_str()));

	  const char* save_format="PPM";
	  if (save_filename.toUpper().endsWith(".PPM"))
	    {
	      save_format="PPM";
	    }
	  else if (save_filename.toUpper().endsWith(".PNG"))
	    {
	      save_format="PNG";
	    }
	  else
	    {
	      std::cerr 
		<< "evolvotron_render: Warning: Unrecognised file suffix.  File will be written in "
		<< save_format
		<< " format.\n";
	    }

	  if (frames>1)
	    {
	      QString frame_component;
	      frame_component.sprintf(".f%06d",frame);
	      int insert_point=save_filename.lastIndexOf(QString("."));
	      if (insert_point==-1)
		{
		  save_filename.append(frame_component);
		}
	      else
		{
		  save_filename.insert(insert_point,frame_component);
		}
	    }
    
	  QImage image
	    (
	     reinterpret_cast<uchar*>(&(image_data[0])),
	     width,
	     height,
	     QImage::Format_RGB32
	     );

	  if (!image.save(save_filename,save_format))
	    {
	      std::cerr 
		<< "evolvotron_render: Error: Couldn't save file "
		<< save_filename.toLocal8Bit().data()
		<< "\n";
	      return 1;
	    }
	
	  std::clog
	    << "Wrote file " 
	    << save_filename.toLocal8Bit().data()
	    << "\n";
	}
      }
  }
  
  return 0;
}
